AlpsSDK/ProximityHandler.swift

```diff
@@ -0,0 +1,136 @@
+//  AlpsSDK
+//
+//  Created by Wen on 26.10.17.
+//  Copyright © 2017 Alps. All rights reserved.
+//
+
+import Foundation
+import CoreLocation
+import Alps
+
+protocol ProximityHandlerDelegate : class {
+    func didRangeBeacons(manager: ContextManager, beacons: [CLBeacon], knownBeacons: [IBeaconTriple])
+}
+
+class ProximityHandler: NSObject, ProximityHandlerDelegate {
+    var refreshTimer: Int = 60 * 1000 // timer is in milliseconds
+    var beaconsDetected: [String: [IBeaconTriple]] = [:]
+    var beaconsTriggered: [String: [String: ProximityEvent]] = [:]
+    
+    override init() {
+        super.init()
+        beaconsTriggered["unknown"] = [String: ProximityEvent]()
+        beaconsTriggered["immediate"] = [String: ProximityEvent]()
+        beaconsTriggered["far"] = [String: ProximityEvent]()
+        beaconsTriggered["near"] = [String: ProximityEvent]()
+    }
+    
+    func didRangeBeacons(manager: ContextManager, beacons: [CLBeacon], knownBeacons: [IBeaconTriple]) {
+        let unknownBeacons = beacons.filter {$0.proximity == CLProximity.unknown}
+        let immediateBeacons = beacons.filter {$0.proximity == CLProximity.immediate}
+        let nearBeacons = beacons.filter {$0.proximity == CLProximity.near}
+        let farBeacons = beacons.filter {$0.proximity == CLProximity.far}
+        
+        // Filter with existing beacons in application
+        beaconsDetected["unknown"] = synchronizeBeacons(beacons: unknownBeacons, knownBeacons: knownBeacons)
+        beaconsDetected["immediate"] = synchronizeBeacons(beacons: immediateBeacons, knownBeacons: knownBeacons)
+        beaconsDetected["near"] = synchronizeBeacons(beacons: nearBeacons, knownBeacons: knownBeacons)
+        beaconsDetected["far"] = synchronizeBeacons(beacons: farBeacons, knownBeacons: knownBeacons)
+        
+        // triggers for proximity event
+        beaconsDetected.forEach { (key, value) in
+            handleProximity(key: key, value: value)
+        }
+    }
+    
+    private func handleProximity(key: String, value: [IBeaconTriple]) {
+        let distance = setUp(key: key)
+        value.forEach { (iBeaconTriple) in
+            guard let deviceId = iBeaconTriple.deviceId else {return}
+            guard let keys = beaconsTriggered[key]?.keys else {return}
+            if keys.contains(deviceId) {
+                // do refresh
+                refreshTriggers(key: key, deviceId: deviceId, distance: distance)
+            } else {
+                // do trigger
+                triggers(key: key, deviceId: deviceId, distance: distance)
+            }
+        }
+    }
+    
+    // Trigger the proximity event
+    private func triggers(key: String, deviceId: String, distance: Double) {
+        let proximityEvent = ProximityEvent.init(deviceId: deviceId, distance: distance)
+        sendProximityEvent(deviceId: deviceId, proximityEvent: proximityEvent) {(_ proximityEvent) in
+            guard let pe = proximityEvent else {return}
```

Comment:
`guard let proxmityEvent = proximityEvent else { return }` seem to be correct approach for now.

I hate it, but swift community says - that's how we should roll :P

Reply:
haha yes, we could think of implementing sort of Error handler. TBD